module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -780,243 +804,110 @@ namespace app
-            auto popupParams = static_cast<const gui::PhoneModePopupRequestParams *>(params);
-            handlePhoneModeChanged(popupParams->getPhoneMode());
-        }
-        else if (id == ID::Volume) {
-            auto volumeParams = static_cast<const gui::VolumePopupRequestParams *>(params);
-            LOG_INFO("Playback: %s, volume: %s",
-                     audio::str(volumeParams->getAudioContext().second).c_str(),
-                     std::to_string(volumeParams->getVolume()).c_str());
-            handleVolumeChanged(volumeParams->getVolume(), volumeParams->getAudioContext());
-        }
-        else if (id == ID::PhoneLock) {
-            switchWindow(
-                gui::popup::resolveWindowName(id), gui::ShowMode::GUI_SHOW_INIT, nullptr, SwitchReason::PhoneLock);
-        }
-        else if (id == ID::PhoneLockInput || id == ID::PhoneLockChangeInfo) {
-            auto popupParams = static_cast<const gui::PhoneUnlockInputRequestParams *>(params);
-
-            switchWindow(
-                gui::popup::resolveWindowName(id),
-                std::make_unique<locks::LockData>(popupParams->getLock(), popupParams->getPhoneLockInputTypeAction()));
-        }
-        else if (id == ID::SimLock || id == ID::SimInfo) {
-            auto popupParams = static_cast<const gui::SimUnlockInputRequestParams *>(params);
-
-            switchWindow(gui::popup::resolveWindowName(id),
-                         std::make_unique<locks::SimLockData>(popupParams->getLock(),
-                                                              popupParams->getSimInputTypeAction(),
-                                                              popupParams->getErrorCode()));
-        }
-        else if (id == ID::Alarm) {
-            auto popupParams =
-                const_cast<gui::AlarmPopupRequestParams *>(static_cast<const gui::AlarmPopupRequestParams *>(params));
-            switchWindow(gui::popup::resolveWindowName(id),
-                         std::make_unique<gui::AlarmPopupRequestParams>(popupParams));
-        }
-        else {
-            switchWindow(gui::popup::resolveWindowName(id), gui::ShowMode::GUI_SHOW_INIT, nullptr, SwitchReason::Popup);
+        auto request = windowsPopupQueue->popRequest(getPopupFilter());
+        if (request) {
+            auto popup = std::string(magic_enum::enum_name(request->getPopupParams().getPopupId()));
+            LOG_DEBUG("handling popup: %s", popup.c_str());
+            /// request handle actually switches window to popup window
+            auto retval = request->handle();
+            if (not retval) {
+                LOG_ERROR("Popup %s handling failure, please check registered blueprint!", popup.c_str());
+            }
+            return retval;
+            LOG_ERROR("no blueprint for popup: %s", popup.c_str());
         }
+        return false;
     }
 
+    // oh this is kind of bad... we require popup start window to properly abort selected popup XD
```

Comment:
rem

Reply:
The popup abort actually closes the popup we want to. So it is desired to start the popup. It's not meant to only delete it from popups stack

Reply:
generally speaking, our popups integrated into the windows stack was and is just a quick and nifty trick o add popups to our already existing window stack. I would rather say that what we have here is the aftermath.
Continuing - while it was added I opted in to create another window stack on top - just to have full popup context for the time we are actually operating the popup. That would add more granular control, make flows more distinguishable and finally enable us in the future to easily have a popup for part of the screen hovering over the screen shown. 
We have what we have, but the comment shall be removed :+1: